File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit_objc/FIAPReceiptManager.m
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "./include/in_app_purchase_storekit_objc/FIAPReceiptManager.h"

#if TARGET_OS_OSX
#import <FlutterMacOS/FlutterMacOS.h>
#else
#import <Flutter/Flutter.h>
#endif
#import "./include/in_app_purchase_storekit_objc/FIAObjectTranslator.h"

@interface FIAPReceiptManager ()
// Gets the receipt file data from the location of the url. Can be nil if
// there is an error. This interface is defined so it can be stubbed for testing.
- (NSData *)getReceiptData:(NSURL *)url error:(NSError **)error;
// Gets the app store receipt url. Can be nil if
// there is an error. This property is defined so it can be stubbed for testing.
@property(nonatomic, readonly) NSURL *receiptURL;
@end

@implementation FIAPReceiptManager

- (NSString *)retrieveReceiptWithError:(FlutterError **)flutterError {
  NSURL *receiptURL = self.receiptURL;
  if (!receiptURL) {
    return nil;
  }
  NSError *receiptError;
  NSData *receipt = [self getReceiptData:receiptURL error:&receiptError];
  if (!receipt || receiptError) {
    if (flutterError) {
      NSDictionary *errorMap = [FIAObjectTranslator getMapFromNSError:receiptError];
      *flutterError =
          [FlutterError errorWithCode:[NSString stringWithFormat:@"%@", errorMap[@"code"]]
                              message:errorMap[@"domain"]
                              details:errorMap[@"userInfo"]];
    }
    return nil;
  }
  return [receipt base64EncodedStringWithOptions:kNilOptions];
}

- (NSData *)getReceiptData:(NSURL *)url error:(NSError **)error {
  return [NSData dataWithContentsOfURL:url options:NSDataReadingMappedIfSafe error:error];
}

- (NSURL *)receiptURL {
  return [[NSBundle mainBundle] appStoreReceiptURL];
}

@end

File: packages/in_app_purchase/in_app_purchase_storekit/darwin/in_app_purchase_storekit/Sources/in_app_purchase_storekit/FIAPReceiptManager.swift
// Copyright 2013 The Flutter Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

#if os(iOS)
  import Flutter
#elseif os(macOS)
  import FlutterMacOS
#endif

public class FIAPReceiptManager: NSObject {
  public override init() {}

  public func retrieveReceiptWithError(_ flutterError: inout FlutterError?) -> String? {
    guard let receiptURL = receiptURL else {
      return nil
    }
    var receiptError: NSError?
    let receipt = getReceiptData(receiptURL, error: &receiptError)
    guard let receipt = receipt, receiptError == nil else {
      let errorMap = FIAObjectTranslator.getMapFrom(receiptError ?? NSError())
      flutterError = FlutterError(
        code: "\(errorMap["code"] ?? "")",
        message: errorMap["domain"] as? String,
        details: errorMap["userInfo"])
      return nil
    }
    return receipt.base64EncodedString()
  }

  /// Gets the receipt file data from the location of the url. Can be nil if
  /// there is an error. This method is defined so it can be overridden for testing.
  @objc(getReceiptData:error:)
  public func getReceiptData(_ url: URL, error: NSErrorPointer) -> Data? {
    do {
      return try Data(contentsOf: url, options: .mappedIfSafe)
    } catch let dataError as NSError {
      error?.pointee = dataError
      return nil
    }
  }

  /// Gets the app store receipt url. Can be nil if
  /// there is an error. This property is defined so it can be overridden for testing.
  @objc public var receiptURL: URL? {
    return Bundle.main.appStoreReceiptURL
  }
}
